File: examples/spectral/static_thermal/cpu_sz.cpp
// =============================================================================
// examples/spectral/static_thermal/cpu_sz.cpp
//
// spectral | static thermal | CPU (OpenMP) | U(1)-Sz, half-filled (Sz=0, n_up=N/2)
//
// static thermodynamic averages at a single T via qed.thermal. Twin: examples/spectral/static_thermal/cpu_sz.py
//
// Requires: (no special requirements; runs on the default CPU build)
// =============================================================================

#include "../../_shared/common.h"

#include <ed/api.h>

#include <iomanip>
#include <iostream>

int main(int argc, char** argv) {
    auto guard = ed_example::mpi_guard(argc, argv);
    constexpr std::uint64_t N = 8;
    (void)guard;

    auto op   = ed_example::heisenberg_chain(N, /*pbc=*/true);
    auto spec = ed_example::in_memory_spec(std::move(op), N);

    ed::api::ThermalOptions opts;
    opts.method      = "FTLM";
    opts.device      = "cpu";
    opts.T_min       = 0.5;
    opts.T_max       = 2.0;
    opts.num_T       = 4;
    opts.num_samples = 8;
    opts.random_seed = 0;
    opts.use_sz_if_conserved = true;
    opts.sz_min = static_cast<int>(N / 2);
    opts.sz_max = static_cast<int>(N / 2);

    auto result = ed::api::thermal(std::move(spec), opts);

    const auto& T  = result.thermo.temperatures;
    const auto& E  = result.thermo.energy;
    const auto& Cv = result.thermo.specific_heat;

    std::cout << std::fixed << std::setprecision(4);
    ed_example::rank0_print("gs_E    = ", result.ground_state_energy, "\n");
    if (!T.empty()) {
        const std::size_t iT = T.size() / 2;
        ed_example::rank0_print(
            "T_probe = ", T[iT], "  E = ", E[iT], "  Cv = ", Cv[iT], "\n");
    }

    // === Expected output (deterministic; captured on the CI reference runner) ===
    // gs_E    = -2.8204
    // T_probe = 1.5000  E = -1.3940  Cv = 1.0370
    // ===========================================================================
    return 0;
}

File: examples/thermal/ftlm/cpu_sz.py
"""thermal | FTLM | CPU (OpenMP) | U(1)-Sz auto-decomposition

Finite-Temperature Lanczos: random vectors x Lanczos. Twin: ``examples/thermal/ftlm/cpu_sz.cpp``.

Requires: (no special requirements; runs on the default CPU build)
"""
from __future__ import annotations

import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", ".."))
from _shared.common import heisenberg_chain, rank0_print

import qed

N = 8
H = heisenberg_chain(N, pbc=True)

result = qed.thermal(
    H,
    method="FTLM",
    T_min=0.1,
    T_max=10.0,
    num_T=8,
    num_samples=8,
    random_seed=0,
    device="cpu",
    use_sz_if_conserved=True,
    verbose=False,
)

T  = result.temperatures
E  = result.energy
Cv = result.specific_heat
mid = len(T) // 2

rank0_print(f"gs_E    = {result.ground_state_energy:.4f}")
rank0_print(f"T[0]    = {T[0]:.4f}  E = {E[0]:.4f}  Cv = {Cv[0]:.4f}")
rank0_print(f"T[mid]  = {T[mid]:.4f}  E = {E[mid]:.4f}  Cv = {Cv[mid]:.4f}")
rank0_print(f"T[-1]   = {T[-1]:.4f}  E = {E[-1]:.4f}  Cv = {Cv[-1]:.4f}")

# === Expected output (deterministic; captured on the CI reference runner) ===
# gs_E    = -3.6484
# T[0]    = 0.1000  E = -3.6420  Cv = 0.4751
# T[mid]  = 5.7571  E = -1.0500  Cv = 0.6946
# T[-1]   = 10.0000  E = -0.1798  Cv = 0.0161
# ===========================================================================
